contracts/buy/PartyBuy.sol

```diff
@@ -0,0 +1,180 @@
+            require(_splitBasisPoints == 0, "PartyBuy::initialize: can't send tokens to burn addr");
+        }
+        splitBasisPoints = _splitBasisPoints;
+        splitRecipient = _splitRecipient;
+        // validate token exists
+        require(_getOwner() != address(0), "PartyBuy::initialize: NFT getOwner failed");
+    }
+
+    // ======== External: Contribute =========
+
+    /**
+     * @notice Contribute to the Party's treasury
+     * while the Party is still active
+     * @dev Emits a Contributed event upon success; callable by anyone
+     */
+    function contribute() external payable nonReentrant {
+        // require that the new total contributed is not greater than
+        // the maximum amount the Party is willing to spend
+        require(totalContributedToParty + msg.value <= getMaximumContributions(), "PartyBuy::contribute: cannot contribute more than max");
+        // continue with shared _contribute flow
+        _contribute();
+    }
+
+    // ======== External: Buy =========
+
+    /**
+     * @notice Buy the token by calling targetContract with calldata supplying value
+     * @dev Emits a Bought event upon success; reverts otherwise. callable by anyone
+     */
+    function buy(uint256 _value, address _targetContract, bytes calldata _calldata) external nonReentrant {
+        require(
+            partyStatus == PartyStatus.ACTIVE,
+            "PartyBuy::buy: party not active"
+        );
+        // check that value is not more than the maximum price set at deploy time
+        require(_value <= maxPrice, "PartyBuy::buy: can't spend over max price");
+        // check that value is not more than
+        // the maximum amount the party can spend while paying ETH fee
+        require(_value <= getMaximumSpend(), "PartyBuy::buy: insuffucient funds to buy token plus fee");
+        // execute the calldata on the target contract
+        address(_targetContract).call{value: _value}(_calldata);
+        // NOTE: we don't are if the call succeeded
+        // as long as the NFT is owned by the Party
+        require(_getOwner() == address(this), "PartyBuy::buy: failed to buy token");
+        // set partyStatus to WON
+        partyStatus = PartyStatus.WON;
+        // record totalSpent,
+        // send ETH fees to PartyDAO,
+        // fractionalize the Token
+        // send Token fees to PartyDAO & split proceeds to split recipient
+        uint256 _ethFee = _closeSuccessfulParty(_value);
+        // emit Bought event
+        emit Bought(msg.sender, _targetContract, _value, _ethFee, totalContributedToParty);
+    }
+
+    // ======== External: Fail =========
+
+    /**
+     * @notice If the token couldn't be successfully bought
+      * within the specified period of time, move to FAILED state
+      * so users can reclaim their funds.
+     * @dev Emits a Closed event upon finishing; reverts otherwise.
+     * callable by anyone after expiresAt
+     */
+    function closeParty() external {
```

Comment:
`closeParty` sounds like n action that could be similar to `finalize`.  would `expireParty` be more of a suiting name?

Reply:
super open to any suggestions, i was stuck on this naming tbh

Reply:
i prefer `expireParty` 👍

Reply:
good call also because I am using `_closeSuccessfulParty` for an internal function when the Party wins